### ISSUE TITLE
Write requires await to work as in Express.js

### ISSUE THREAD

**Issue by skakac** [kartikk221/hyper-express#256]
Im trying out hyper-express so I can test if I can switch to it in a future refactor.

What I found is that if I have code like this:

```
webserver.get('/', (request: Request, response: Response) => {
  response.write('1');
  response.write('2');
  response.send('done');
});
```
Output will be: `1done`

I need to add await to .write calls if I want all of them to execute before .send.

```
webserver.get('/', async (request: Request, response: Response) => {
  await response.write('1');
  await response.write('2');
  response.send('done');
});
```
Output will be: `12done`

I could use callback argument of .write, but it's ugly and not really express like.

I get that this is something to do with uWS and maybe corking, so if its hard to make .write sync function, maybe adding something like:

`write(chunk: any, callback?: (error: Error | null | undefined) => void): Promise<boolean>;`
`write(chunk: any, encoding: BufferEncoding, callback?: (error: Error | null | undefined) => void): Promise<boolean>;`

to `export class Response<Locals = DefaultResponseLocals> extends Writable {` will do as a quick fix, so IDEs and users will know it should be used as async function.

**Comment by kartikk221** [comment]
Hey, good catch and yeah writing is async (regardless of Express allowing sync, they are likely just queueing it).
Do you want to make a PR with those updated TypeScript definitions?

**Comment by doxlik** [comment]
Sorry, but how can these methods be defined while actually they don't have implementation? I mean we do not have any **"write"** method implementation in the Response class. It has implementation of the **"_write"** method which then will be automatically called by the Writable parent object when we call its **"write"**. And so there is no meaning of defining any custom definitions for the **"write"** methods, as it can cause the bugs when the actual **"write"** method will return boolean but we may try to call **"then"** on it as we expect promise and then error will be thrown as we cannot call **"then"** method on the boolean. Please correct me if I am mistaken.

**Comment by kartikk221** [comment]
So `Response` is actually an extended `WritableStream` (We do a trick to merge prototypes to lazy initialize the stream when used https://github.com/kartikk221/hyper-express/blob/e98212ee8bb4273b6cf8c6b025d9c52e15cc9df0/src/components/http/Response.js#L958-L983) hence the `Response.write()` method being called is actually internally calling the `WritableStream` write method which then internally calls `Response._write()`.

The implementation of `Response._write()` respects the arguments and properly calls the `callback` (if provided) which would mean Node.js should property queue and release the written chunks properly.

Now that I think about it, I think your code may actually be pointing to a bug where `Response.send()` should understand that there is currently a stream write in progress hence it must queue the provided data as a chunk and then finally send the response once the stream has finished.

I will do some debugging into this to see If it is a bug which must be fixed.

**Comment by kartikk221** [comment]
Hey, this bug has been fixed in the recent `v6.16.1` published version. Let me know If you have any more issues.
